### PR TITLE
Improve `-Zhir-stats`

### PR DESCRIPTION
**Pull request by nnethercote**
Add testing, improve coverage, avoid some double counting, and add more detail.

r? @michaelwoerister

### CODE REVIEW

File: compiler/rustc_passes/src/hir_stats.rs [hunk 1]
```diff
@@ -268,14 +320,62 @@ impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
...
+            $(
+                ast::$tykind::$variant(..) => {
+                    $self.record_variant(stringify!($ty), stringify!($variant), Id::None, $val)
+                }
+            )*
+            $(
+                ast::$tykind::$variant0 => {
+                    $self.record_variant(stringify!($ty), stringify!($variant0), Id::None, $val)
+                }
+            )*
```
Comment (michaelwoerister) [inline compiler/rustc_passes/src/hir_stats.rs:340]: I think `ast::$tykind::$variant {..}` would work for all cases:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=09ab4aa5b8ab740d8ee4a59b721aff8f

**Review by michaelwoerister — COMMENTED** [review]
Thanks for the PR, @nnethercote! This looks useful indeed.

I left a comment about a possible simplification below. Otherwise, feel free to r=me.

File: compiler/rustc_passes/src/hir_stats.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_passes/src/hir_stats.rs:340]: Nice! I was wondering if there was some trick like that. I'll do it tomorrow.

**Comment by nnethercote** [comment]
I made the suggested simplification.

@bors r=michaelwoerister

**Comment by nnethercote** [comment]
@bors rollup=always

Because these code changes are all behind the `-Zhir-stats` option, so don't affect normal execution.

**Pull request merged**